core/storage/email/gae_models_test.py

```diff
@@ -30,9 +30,94 @@ def test_sent_email_model_instances_are_read_only(self):
         email_models.SentEmailModel.create(
             'recipient_id', 'recipient@email.com', 'sender_id',
             'sender@email.com', feconf.EMAIL_INTENT_SIGNUP,
-            'Email Subject', 'Email Body', datetime.datetime.utcnow())
+            'Email Subject', 'Email Body', datetime.datetime.utcnow(),
+            'Email Hash')
 
         model = email_models.SentEmailModel.get_all().fetch()[0]
         model.recipient_id = 'new_recipient_id'
         with self.assertRaises(Exception):
             model.put()
+
+    def test_model_saved_with_certain_hash_can_be_retrieved_same_hash(self):
+        email_models.SentEmailModel.create(
+            'recipient_id', 'recipient@email.com', 'sender_id',
+            'sender@email.com', feconf.EMAIL_INTENT_SIGNUP,
+            'Email Subject', 'Email Body', datetime.datetime.utcnow(),
+            'Email Hash')
+
+        query = email_models.SentEmailModel.query()
+        query = query.filter(
+            email_models.SentEmailModel.email_hash == 'Email Hash')
+
+        results = query.fetch(2)
+
+        self.assertEqual(len(results), 1)
+
+        query = email_models.SentEmailModel.query()
+        query = query.filter(
+            email_models.SentEmailModel.email_hash == 'Wrong Email Hash')
+
+        results = query.fetch(2)
+
+        self.assertEqual(len(results), 0)
+
+    def test_model_saved_certain_hash_retrieved_same_hash_using_getbyhash(self):
+        email_models.SentEmailModel.create(
+            'recipient_id', 'recipient@email.com', 'sender_id',
+            'sender@email.com', feconf.EMAIL_INTENT_SIGNUP,
+            'Email Subject', 'Email Body', datetime.datetime.utcnow(),
+            'Email Hash')
+
+        results = email_models.SentEmailModel.get_by_hash('Email Hash')
+
+        self.assertEqual(len(results), 1)
+
+        results = email_models.SentEmailModel.get_by_hash('Wrong Email Hash')
+
+        self.assertEqual(len(results), 0)
+
+    def test_get_by_hash_return_multiple_models_same_hash(self):
+        email_models.SentEmailModel.create(
+            'recipient_id', 'recipient@email.com', 'sender_id',
+            'sender@email.com', feconf.EMAIL_INTENT_SIGNUP,
+            'Email Subject', 'Email Body', datetime.datetime.utcnow(),
+            'Same Hash')
+
+        email_models.SentEmailModel.create(
+            'recipient_id', 'recipient@email.com', 'sender_id',
+            'sender@email.com', feconf.EMAIL_INTENT_SIGNUP,
+            'Email Subject', 'Email Body', datetime.datetime.utcnow(),
+            'Same Hash')
+
+        results = email_models.SentEmailModel.get_by_hash('Same Hash')
+
+        self.assertEqual(len(results), 2)
+
+    def test_get_by_hash_correct_usage_after_argument(self):
```

Comment:
It's no longer called 'after'; you may want to change that below.

Suggest: test_get_by_hash_behavior_with_sent_datetime_lower_bound